5496 - Validation bug 12883

speratimatteo opened the pull request:
Aggiunti scenari di test per verificare il comportamento dell'update massivo in caso di file key non effettivamente presenti sul DB

speratimatteo committed:
5496 - Validation bug 12883
Aggiunti scenari di test per verificare il comportamento dell'update massivo in caso di file key non effettivamente presenti sul DB

speratimatteo committed:
5496 - Validation bug 12883
Aggiunti scenari di test per verificare il comportamento dell'update massivo in caso di tag inesistenti

speratimatteo committed:
5496 - Validation bug 12883
Aggiunti scenari di test per verificare il comportamento dell'update massivo in caso di tag inesistenti

speratimatteo committed:
5496 - Validation bug 12883
Aggiunti scenari di test per verificare il comportamento dell'update massivo in caso di tag inesistenti

speratimatteo committed:
5496 - Validation bug 12883
Aggiunti scenari di test per verificare il comportamento dell'update massivo in caso di tag inesistenti

speratimatteo committed:
Merge branch 'develop' into feature/5496-validationBug12883

speratimatteo committed:
Merge branch 'develop' into feature/5496-validationBug12883

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Rework su alcuni vecchi test in modo da poterli far funzionare con i limiti aggiornati (e renderli globalmente più robusti e comprensibili)

speratimatteo committed:
5496 - Validation bug 12883

Da quando su ParameterStore di confinfo sono stati resi uguali i valori dei limiti di Pn-SS-IndexingConfiguration
di TEST e UAT, le due suite che erano state pensate per essere eseguite nei due ambienti non sono più necessarie.
Verrà pertanto utilizzata un'unica suite valevole per entrambi gli ambienti.
Il codice è stato notevolmente irrobustito e reso di più facile fruizione e mantenibilità in caso di update futuri.

I test:
INDEX_SS_CREATE_3
INDEX_SS_CREATE_6
INDEX_SS_UPDATE_SINGLE_8
INDEX_SS_UPDATE_SINGLE_10
INDEX_SS_UPDATE_MASSIVE_9
INDEX_SS_UPDATE_MASSIVE_12

sono stati tenuti, ma presentano l'annotazione @indicizzazioneSafeStorageOLD
che li escluderà pertanto dalla run.
Questo perchè per poter verificare i casi di errore che tali test volevano replicare, sarebbe necessario un numero di tag superiore a quelli attualmente presenti su indexingConfiguration.

mmancini95 commented on src/test/resources/it/pagopa/pn/cucumber/IndicizzazioneFile/IndicizzazioneFile.feature:
```diff
@@ -74,9 +74,9 @@ Feature: test preliminari indicizzazione File safeStorage
-  @test
   @aggiuntaTag
-  @indicizzazioneSafeStorage
+  @indicizzazioneSafeStorageOLD
```
è possibile raggruppare i vecchi test in un feature file dedicato in modo da tenere questo pulito e lasciare gli altri solo per tracciabilità?

mmancini95 commented on src/test/resources/it/pagopa/pn/cucumber/IndicizzazioneFile/IndicizzazioneFile.feature:
```diff
@@ -206,21 +195,21 @@ Feature: test preliminari indicizzazione File safeStorage
-  @test
   @aggiuntaTag
-  @indicizzazioneSafeStorage
+  @indicizzazioneSafeStorageTODO
```
è possibile creare un ticket tecnico per questi test marcati con: `@indicizzazioneSafeStorageTODO` in modo da non perdere traccia di questi test da rivedere

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/utils/IndicizzazioneStepsPojo.java:
```diff
@@ -1,30 +1,38 @@
-import it.pagopa.pn.client.web.generated.openapi.clients.safeStorage.model.FileCreationResponse;
-import java.util.ArrayList;
-import java.util.List;
+import it.pagopa.pn.client.web.generated.openapi.clients.safeStorage.model.*;
```
è possibile rimuovere questo wildcard import e usare gli import specifici?

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/microservice/SafeStorageSteps.java:
```diff
@@ -43,6 +47,219 @@ private String computeSha(String resourceName) {
+        }
+    }
+
+    private Integer eseguiGetterDelLimite(IndicizzazioneStepsPojo pojo, String fieldName) {
```
```suggestion
    private Integer retrieveLimitFromPojo(IndicizzazioneStepsPojo pojo, String fieldName) {
```

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/microservice/SafeStorageSteps.java:
```diff
@@ -43,6 +47,219 @@ private String computeSha(String resourceName) {
+    private int getLimitValue(String comparator, String limit) {
+        int quantity;
+        try {
+            quantity = Integer.parseInt(limit);
```
Dato che la maggior parte dei metodi che richiamano questo step gli passano un limit che è una stringa che non rappresenta un intero, non sarebbe meglio lasciare in questo metodo solo la parte all'interno del catch?
Gli step invece in cui viene passato un intero potrebbero essere cambiati in modo da ottenere direttamente un valore int invece che una stringa così non è necessario il parseInt, vedi commento sopra

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/microservice/SafeStorageSteps.java:
```diff
@@ -43,6 +47,219 @@ private String computeSha(String resourceName) {
+        uploadDocumentsWithTags(type, tagMap, quantity);
+    }
+
+    @Given("vengono caricati documenti di tipo {string} in numero {string} a {string} con associato il tag {string} avente {int} valori diversi")
+    public void uploadMultipleDocumentsWithAssociatedTagsWithValues(String type, String comparator, String limit, String tagName, Integer valueNumber) {
```
```suggestion
    @Given("vengono caricati documenti di tipo {string} in numero {string} a {int} con associato il tag {string} avente {int} valori diversi")
    public void uploadMultipleDocumentsWithAssociatedTagsWithValues(String type, String comparator, int limit, String tagName, Integer valueNumber) {
```

mmancini95 commented on src/test/java/it/pagopa/pn/cucumber/steps/microservice/SafeStorageSteps.java:
```diff
@@ -43,6 +47,219 @@ private String computeSha(String resourceName) {
+        uploadDocumentsWithTags(type, tagMap, quantity);
+    }
+
+    @Given("il documento viene aggiornato aggiungendo {string} valori per volta al tag {string}, fino a raggiungere il limite di {string}")
+    public void singleAddValuesUntilMax(String maxValuesPerTagPerRequest, String tagName, String maxValuesPerTagDocument) {
+        int maxValuesPerTagPerRequestInt = eseguiGetterDelLimite(this.indicizzazioneStepsPojo, maxValuesPerTagPerRequest);
+        int maxValuesPerTagDocumentInt = eseguiGetterDelLimite(this.indicizzazioneStepsPojo, maxValuesPerTagDocument);
+        int counterTagsAdded = 0;
+        while (counterTagsAdded < maxValuesPerTagDocumentInt) {
+            int valuesToAdd = getQuantityToAddInIteration(maxValuesPerTagDocumentInt, maxValuesPerTagPerRequestInt, counterTagsAdded);
+            updateSingleContinuativo(counterTagsAdded, tagName, "PARI", String.valueOf(valuesToAdd));
+            counterTagsAdded += valuesToAdd;
+        }
+    }
+
+    @Given("i documenti vengono aggiornati aggiungendo {string} valori per volta al tag {string}, fino a raggiungere il limite di {string}")
+    public void multipleAddValuesUntilMax(String maxValuesPerTagPerRequest, String tagName, String maxValuesPerTagDocument) {
+        int maxValuesPerTagPerRequestInt = eseguiGetterDelLimite(this.indicizzazioneStepsPojo, maxValuesPerTagPerRequest);
+        int maxValuesPerTagDocumentInt = eseguiGetterDelLimite(this.indicizzazioneStepsPojo, maxValuesPerTagDocument);
+        int counterTagsAdded = 0;
+        while (counterTagsAdded < maxValuesPerTagDocumentInt) {
+            int valuesToAdd = getQuantityToAddInIteration(maxValuesPerTagDocumentInt, maxValuesPerTagPerRequestInt, counterTagsAdded);
+            updateMassiveContinuativo(counterTagsAdded, tagName, "PARI", String.valueOf(valuesToAdd));
+            counterTagsAdded += valuesToAdd;
+        }
+    }
```
per evitare questa duplicazione di codice si potrebbe attuare una soluzione del genere:
```suggestion
    @FunctionalInterface
    public interface QuadConsumer<T, U, V, W> {
        void accept(T t, U u, V v, W w);
    }

    @Given("il documento viene aggiornato aggiungendo {string} valori per volta al tag {string}, fino a raggiungere il limite di {string}")
    public void singleAddValuesUntilMax(String maxValuesPerTagPerRequest, String tagName, String maxValuesPerTagDocument) {
        addValueUntilMax(maxValuesPerTagPerRequest, tagName, maxValuesPerTagDocument, this::updateSingleContinuativo);
    }

    @Given("i documenti vengono aggiornati aggiungendo {string} valori per volta al tag {string}, fino a raggiungere il limite di {string}")
    public void multipleAddValuesUntilMax(String maxValuesPerTagPerRequest, String tagName, String maxValuesPerTagDocument) {
        addValueUntilMax(maxValuesPerTagPerRequest, tagName, maxValuesPerTagDocument, this::updateMassiveContinuativo);
    }

    private void addValueUntilMax(String maxValuesPerTagPerRequest, String tagName, String maxValuesPerTagDocument, QuadConsumer<Integer, String, String, String> updateMethod) {
        int maxValuesPerTagPerRequestInt = eseguiGetterDelLimite(this.indicizzazioneStepsPojo, maxValuesPerTagPerRequest);
        int maxValuesPerTagDocumentInt = eseguiGetterDelLimite(this.indicizzazioneStepsPojo, maxValuesPerTagDocument);
        int counterTagsAdded = 0;
        while (counterTagsAdded < maxValuesPerTagDocumentInt) {
            int valuesToAdd = getQuantityToAddInIteration(maxValuesPerTagDocumentInt, maxValuesPerTagPerRequestInt, counterTagsAdded);
            updateMethod.accept(counterTagsAdded, tagName, "PARI", String.valueOf(valuesToAdd));
            counterTagsAdded += valuesToAdd;
        }
    }
```

speratimatteo committed:
5496 - Validation bug 12883

Modifiche in seguito a review di Matteo Mancini.
- Codice duplicato rimosso
- Evitato uso del ParseInt con cattura dell'eccezione.
- Vecchi test spostati in un file feature a parte
